ouroboros-network/src/Ouroboros/Cardano/PeerSelection/Churn.hs

```diff
@@ -0,0 +1,601 @@
+                    (1 + closeConnectionTimeout)
+                    decreaseEstablishedBigLedgerPeers
+                    checkEstablishedBigLedgerPeersDecreased
+
+      -- Forget the worst performing known peers (root peers, ledger peers)
+      updateTargets DecreasedKnownPeers
+                    numberOfKnownPeers
+                    shortTimeout
+                    decreaseKnownPeers
+                    checkKnownPeersDecreased
+
+      -- Pick new known peers
+      updateTargets IncreasedKnownPeers
+                    numberOfKnownPeers
+                    (2 * requestPeersTimeout + shortTimeout)
+                    increaseKnownPeers
+                    checkKnownPeersIncreased
+
+      -- Forget the worst performing known big ledger peers.
+      updateTargets DecreasedKnownBigLedgerPeers
+                    numberOfKnownBigLedgerPeers
+                    shortTimeout
+                    decreaseKnownBigLedgerPeers
+                    checkKnownBigLedgerPeersDecreased
+
+      -- Pick new known big ledger peers
+      updateTargets IncreasedKnownBigLedgerPeers
+                    numberOfKnownBigLedgerPeers
+                    (2 * requestPeersTimeout + shortTimeout)
+                    increaseKnownBigLedgerPeers
+                    checkKnownBigLedgerPeersIncreased
+
+      -- Pick new non-active peers
+      updateTargets IncreasedEstablishedPeers
+                    numberOfEstablishedPeers
+                    churnEstablishConnectionTimeout
+                    increaseEstablishedPeers
+                    checkEstablishedPeersIncreased
+
+      -- Pick new non-active big ledger peers
+      updateTargets IncreasedEstablishedBigLedgerPeers
+                    numberOfEstablishedBigLedgerPeers
+                    churnEstablishConnectionTimeout
+                    increaseEstablishedBigLedgerPeers
+                    checkEstablishedBigLedgerPeersIncreased
+
+      endTs <- getMonotonicTime
+
+      fuzzyDelay rng (endTs `diffTime` startTs) >>= churnLoop
+
+    --
+    -- Auxiliary functions and constants
+    --
+
+    -- Randomly delay between churnInterval and churnInterval + maxFuzz seconds.
+    fuzzyDelay :: StdGen -> DiffTime -> m StdGen
+    fuzzyDelay rng execTime = do
+      mode <- atomically getFetchMode
+      -- todo: is this right?
+      case (mode, consensusMode) of
+        (PraosFetchMode FetchModeDeadline, _) -> longDelay rng execTime
+        (_, GenesisMode)                      -> longDelay rng execTime
+        _otherwise                            -> shortDelay rng execTime
+
+    fuzzyDelay' :: DiffTime -> Double -> StdGen -> DiffTime -> m StdGen
+    fuzzyDelay' baseDelay maxFuzz rng execTime = do
+      let (fuzz, rng') = randomR (0, maxFuzz) rng
+          delay = realToFrac fuzz + baseDelay - execTime
+      traceWith tracer $ TraceChurnWait delay
+      threadDelay delay
+      return rng'
+
+    longDelay :: StdGen -> DiffTime -> m StdGen
```

Comment:
In the original we have two empty lines separating all the definitions (that was on purpose: one line for things internal to a local binding, two for top level or things in a `where` clause).
```suggestion
    
    longDelay :: StdGen -> DiffTime -> m StdGen
```